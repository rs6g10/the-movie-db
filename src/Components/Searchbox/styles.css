.search-box {
    display: inline-table;
    padding: 5px;
    width: 100%;
}

.search-box_content {
    position: relative;
}

.search-box_input {
    line-height: 30px;
    width: 100%;
    border: 1px solid #01d277;
    border-radius: 2px;
}

.search-box.open .search-box_input {
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
}

.search-box.open .search-box_dropdown {
    display: block;
    z-index: 10;
    opacity: 0.75;
}

.search-box .search-box_dropdown {
    display: none;
    border: 1px solid #01d277;
    background-color: white;
    max-height: 200px;
    overflow: auto;
    border-top: none;
    border-bottom-left-radius: 4px;
    border-bottom-right-radius: 4px;
    box-shadow: 0 2px 1px -2px;
    position: absolute;
    right: 0;
    left: 0;
}

.search-box_dropdown ul {
    padding: 0;
    list-style-type: none;
    margin: 0;
}

.search-box_dropdown ul .search-box_item {
    line-height: 20px;
    padding: 2.5px 5px;
}

.search-box_dropdown ul .search-box_item:hover {
    cursor: pointer;
}

li.search-box_item:first-child {
    border-top: none;
}