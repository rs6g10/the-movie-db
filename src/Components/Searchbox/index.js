import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ClassNames from 'classnames';
import { parseResponse } from '../../utility';

import './styles.css';

class Searchbox extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
      error: null,
      searchResults: null,
      valueInput: '',
    };
    this.onChange = this.onChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.showResults !== nextProps.showResults) {
      this.setState({
        searchResults: nextProps.showResults ? this.state.searchResults : null,
      });
    }
  }

  onChange(event) {
    const value = event.target.value;
    if (value.length > 0) {
      this.fetchData(value);
    } else {
      this.setState({
        searchResults: null,
      });
    }
  }

  displayResultArray(data) {
    return this.generateResults(data);
  }

  fetchData(query) {
    const that = this;
    const url = `https://api.themoviedb.org/3/search/movie?query=${query}&api_key=${this.props.apiKey}`;
    this.setState({
      loading: true,
      error: null,
    });
    fetch(url).then(res => parseResponse(res)).then((data) => {
      if (data && data.results) {
        that.setState({
          searchResults: data.results,
          loading: false,
          error: null,
        });
      }
    }).catch((err) => {
      this.setState({
        error: err,
        loading: false,
      });
    });
  }

  generateResults(data) {
    if (data && data.length > 0) {
      return (
        <ul>
          {
            data.map(datum => (
              <li
                key={datum.id}
                className={
                  ClassNames('search-box_item',
                    { selected: datum.selected },
                    { disable: datum.disable },
                    { hidden: datum.hidden })}
              >
                <Link to={`/movie/${datum.id}`}>
                  {datum.title}
                </Link>
              </li>
            ))
          }
        </ul>
      );
    }
    return null;
  }

  render() {
    const { placeHolder } = this.props;
    const { searchResults, loading, error } = this.state;
    return (
      <div className={ClassNames('search-box', { open: !!searchResults })}>
        <div className="search-box_content">
          <input
            type="text"
            className="search-box_input"
            placeholder={placeHolder}
            onChange={this.onChange}
          />
          <div className="search-box_dropdown">
            {loading && <span>Loading...</span>}
            {!!searchResults && !loading && !error &&
            this.displayResultArray(searchResults)}
            {error &&
            <div className="row">
              <span className="error">{error.message}</span>
            </div>
            }
          </div>
        </div>
      </div>
    );
  }
}

Searchbox.propTypes = {
  apiKey: PropTypes.string.isRequired,
  placeHolder: PropTypes.string,
  showResults: PropTypes.bool,
};

Searchbox.defaultProps = {
  placeHolder: 'Search',
  showResults: true,
};

export default Searchbox;
