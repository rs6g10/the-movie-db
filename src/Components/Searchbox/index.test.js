import '../../test/setupEnzyme';

import React from 'react';
import { shallow } from 'enzyme';

import Searchbox from './';

describe('Header', () => {
  const props = {
    apiKey: 'secret',
    placeHolder: 'Search',
    showResults: true,
  };

  const props2 = {
    apiKey: 'secret',
    placeHolder: 'Search',
    showResults: false,
  };

  it('should render as expected when showResults is set to true', () => {
    const element = shallow(<Searchbox {...props} />);
    expect(element).toMatchSnapshot();
  });

  it('should render as expected when showResults is set to false', () => {
    const element = shallow(<Searchbox {...props2} />);
    expect(element).toMatchSnapshot();
  });
});
