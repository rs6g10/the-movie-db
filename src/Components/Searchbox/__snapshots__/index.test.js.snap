// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render as expected when showResults is set to false 1`] = `
<div
  className="search-box"
>
  <div
    className="search-box_content"
  >
    <input
      className="search-box_input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
    />
    <div
      className="search-box_dropdown"
    />
  </div>
</div>
`;

exports[`Header should render as expected when showResults is set to true 1`] = `
<div
  className="search-box"
>
  <div
    className="search-box_content"
  >
    <input
      className="search-box_input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
    />
    <div
      className="search-box_dropdown"
    />
  </div>
</div>
`;
